lib/jnpr/junos/device.py

```diff
@@ -730,6 +733,30 @@ def facts_refresh(self,
     def __repr__(self):
         return "Device(%s)" % self.hostname
 
+class DeviceSessionListener(SessionListener):
+
+    """
+    Listens to Session class of Netconf Transport and detects errors in the transport.
+    """
+    def __init__(self, device):
+        self._device = device
+    def callback(self, root, raw):
+        """Called when a new XML document is received. The *root* argument allows the callback to determine whether it wants to further process the document.
+
+        Here, *root* is a tuple of *(tag, attributes)* where *tag* is the qualified name of the root element and *attributes* is a dictionary of its attributes (also qualified names).
+
+        *raw* will contain the XML document as a string.
+        """
+        pass
+
+    def errback(self, ex):
+        """Called when an error occurs.
+        Set the device's connected status to False, and then raise a TransportError.
+
+        :type ex: :exc:`Exception`
+        """
+        self._device.connected = False
+        raise NcErrors.TransportError
```

Comment:
I don't believe we should raise an exception here. It's not actually doing anything anyway because ncclient catches the exception and just attempts to log a debug message:
https://github.com/ncclient/ncclient/blob/master/ncclient/transport/session.py#L77-L80

A `TransportError` exception will already be raised the next time there's an attempt to send on this connection:
https://github.com/ncclient/ncclient/blob/master/ncclient/transport/session.py#L155